### ISSUE TITLE
Support all MXNet engine types

### ISSUE THREAD

**Issue by joseph-wakeling-sociomantic** [sociomantic-tsunami/dmxnet#53]
Currently dmxnet only supports `MXNET_ENGINE_TYPE=NaiveEngine`.  However, with libtangort 1.8.x released (and available in the `dlang:v4` base docker image), we should be able to move forward on this.

Trying this out locally it looks like there is some failure in waiting until all calculations complete, since the test suite "works" but generates wrong values:

```
MXNet handles in use at start: 0
[19:15:03] src/engine/engine.cc:36: MXNet start using engine: ThreadedEngine
MXNet handles in use at exit: 0
terminated after throwing an uncaught instance of 'ocean.core.Test.TestException' at integrationtest/mxnet/main.d:146
  toString():  expression '0.10 >= 0.92' evaluates to false
Aborted (core dumped)
```

```
MXNet handles in use at start: 0
[19:15:46] src/engine/engine.cc:36: MXNet start using engine: ThreadedEnginePerDevice
MXNet handles in use at exit: 0
terminated after throwing an uncaught instance of 'ocean.core.Test.TestException' at integrationtest/mxnet/main.d:146
  toString():  expression '0.10 >= 0.92' evaluates to false
Aborted (core dumped)
```

If I remove the failing `test` call, then integration tests with the `ThreadedEngine` report no other failures (the test suite runs and exits successfully), but with `ThreadedEnginePerDevice` the integration tests write their final output and then just hang (the program does not exit).

**Comment by joseph-wakeling-sociomantic** [comment]
Marking as low priority, since while this is important to get at some point, it's not vital to our current plans/use-cases.

**Comment by joseph-wakeling-sociomantic** [comment]
Blocked by https://github.com/sociomantic-tsunami/dmxnet/pull/51, but that should be easy to address.

**Comment by joseph-wakeling-sociomantic** [comment]
We need to update the `libmxnet` package to support multithreading.  This is a necessary precursor before other factors can be addressed, and can be rolled in with our update to use MXNet 1.0.0.

**Comment by joseph-wakeling-sociomantic** [comment]
We now have a libmxnet 1.0.0 package in our prerelease repository.  This should make it possible to start testing with the threaded engines, as IIUC it contains some fixes required on our side.

**Comment by jens-mueller-sociomantic** [comment]
To support threaded engines access to the data of n-dimensional arrays must be properly synchronized. This can be addressed by sociomantic-tsunami/dmxnet#54.
Another related issue is that with the old libmxnet there was a deadlock for the `ThreadedEnginePerDevice`. But the update to `libmxnet 1.0.0` fixed that (very likely fixed by https://github.com/sociomantic-tsunami/mxnet/commit/7b40c031c0147310a5adb0646052508dd4534647#diff-86c84f2855f7be90c3ecff88c583a802). This makes the fixes to the runtime unnecessary. Either one fixes the deadlock issue though.
I think it is safe to move forward with sociomantic-tsunami/dmxnet#54. We should add integrated testing of all engines to get alert for future synchronization or deadlock errors.
